### PR TITLE
[CM-1632] Implement Artifact Get/Download part

### PR DESCRIPTION
**Pull request by yaricom**
### comet-java-client
- Implemented `LoggedArtifactAsset` representing data about asset logged by artifact
- Implemented method to download `LoggedArtifactAsset`'s file from the Comet.
- Implemented method to download `LoggedArtifact` to specific directory.
- Implemented method to load content of `LoggedArtifactAsset` from the Comet server into the memory (ByteBuffer).

### comet-examples
- Implemented example to demonstrate how to download `LoggedArtifactAsset`'s file from the Comet.
- Implemented example to demonstrate how to download `LoggedArtifact` from the Comet.
- Implemented example to demonstrate how to load content of `LoggedArtifactAsset` from the Comet server into the memory (ByteBuffer).

### CODE REVIEW

File: comet-java-client/src/main/java/ml/comet/experiment/impl/BaseArtifactImpl.java [hunk 1]
```diff
@@ -1,25 +1,57 @@
...
+    @Setter
     Set<String> aliases;
-    Map<String, Object> artifactMetadata;
+    @Getter
+    @Setter
     Semver semanticVersion;
+    @Getter
+    @Setter
     Set<String> versionTags;
+    @Setter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/BaseArtifactImpl.java:31]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+/**
+ * Implementation of the {@link LoggedArtifactAsset}.
+ */
+@ToString
+public final class LoggedArtifactAssetImpl implements LoggedArtifactAsset {
+
+    private final Logger logger = LoggerFactory.getLogger(LoggedArtifactAsset.class);
+
+    @Getter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java:33]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    @Setter
+    private String assetId;
+    @Setter
+    private String metadataJson;
+    @Setter
+    @Getter
+    private String fileName;
+    @Setter
+    private Long fileSize;
+    @Setter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java:49]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Implementation of the {@link LoggedArtifactAsset}.
+ */
+@ToString
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java:28]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactImpl.java [hunk 5]
```diff
@@ -1,18 +1,45 @@
...
+import static ml.comet.experiment.impl.resources.LogMessages.FAILED_TO_DOWNLOAD_ARTIFACT_ASSETS;
+import static ml.comet.experiment.impl.resources.LogMessages.START_DOWNLOAD_ARTIFACT_ASSETS;
+import static ml.comet.experiment.impl.resources.LogMessages.getString;
+
 /**
  * The implementation of the {@link LoggedArtifact}.
  */
+@ToString
 public final class LoggedArtifactImpl extends BaseArtifactImpl implements LoggedArtifact {
+    @Getter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactImpl.java:40]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactImpl.java [hunk 6]
```diff
@@ -1,18 +1,45 @@
...
+import static ml.comet.experiment.impl.resources.LogMessages.ARTIFACT_ASSETS_DOWNLOAD_COMPLETED;
+import static ml.comet.experiment.impl.resources.LogMessages.ARTIFACT_HAS_NO_ASSETS_TO_DOWNLOAD;
+import static ml.comet.experiment.impl.resources.LogMessages.FAILED_TO_DOWNLOAD_ARTIFACT_ASSETS;
+import static ml.comet.experiment.impl.resources.LogMessages.START_DOWNLOAD_ARTIFACT_ASSETS;
+import static ml.comet.experiment.impl.resources.LogMessages.getString;
+
 /**
  * The implementation of the {@link LoggedArtifact}.
  */
+@ToString
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactImpl.java:38]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/DownloadArtifactAssetOptions.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+/**
+ * Holds options used to download particular artifact asset.
+ */
+@Data
+@NoArgsConstructor
+@EqualsAndHashCode(callSuper = true)
+@ToString(callSuper = true)
+public class DownloadArtifactAssetOptions extends DownloadAssetOptions {
+    String artifactVersionId;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/DownloadArtifactAssetOptions.java:19]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/DownloadArtifactAssetOptions.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+import lombok.EqualsAndHashCode;
+import lombok.NoArgsConstructor;
+import lombok.ToString;
+
+import java.io.File;
+
+/**
+ * Holds options used to download particular artifact asset.
+ */
+@Data
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/DownloadArtifactAssetOptions.java:14]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/DownloadArtifactAssetOptions.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+@NoArgsConstructor
+@EqualsAndHashCode(callSuper = true)
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/DownloadArtifactAssetOptions.java:16]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/DownloadArtifactAssetOptions.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+@ToString(callSuper = true)
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/DownloadArtifactAssetOptions.java:17]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/DownloadAssetOptions.java [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+/**
+ * Holds options used to download particular artifact asset.
+ */
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class DownloadAssetOptions {
+    String assetId;
+    File file;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/DownloadAssetOptions.java:17]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/DownloadAssetOptions.java [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+import java.io.File;
+
+/**
+ * Holds options used to download particular artifact asset.
+ */
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class DownloadAssetOptions {
+    String assetId;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/DownloadAssetOptions.java:16]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/DownloadAssetOptions.java [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.io.File;
+
+/**
+ * Holds options used to download particular artifact asset.
+ */
+@Data
+@AllArgsConstructor
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/DownloadAssetOptions.java:13]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/DownloadAssetOptions.java [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.io.File;
+
+/**
+ * Holds options used to download particular artifact asset.
+ */
+@Data
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/DownloadAssetOptions.java:12]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/rest/ArtifactVersionAsset.java [hunk 15]
```diff
@@ -17,5 +19,23 @@
     private String dir;
     private String type;
     private String metadata;
-    private Boolean remote;
+    private Boolean remote = Boolean.FALSE;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/rest/ArtifactVersionAsset.java:22]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/rest/ArtifactVersionAssetResponse.java [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.util.List;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class ArtifactVersionAssetResponse {
+    List<ArtifactVersionAsset> files;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/rest/ArtifactVersionAssetResponse.java:13]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/rest/ArtifactVersionAssetResponse.java [hunk 17]
```diff
@@ -0,0 +1,14 @@
+package ml.comet.experiment.impl.rest;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.util.List;
+
+@Data
+@AllArgsConstructor
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/rest/ArtifactVersionAssetResponse.java:10]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/rest/ArtifactVersionAssetResponse.java [hunk 18]
```diff
@@ -0,0 +1,14 @@
+package ml.comet.experiment.impl.rest;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.util.List;
+
+@Data
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/rest/ArtifactVersionAssetResponse.java:9]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/rest/RestApiResponse.java [hunk 19]
```diff
@@ -5,27 +5,38 @@
...
 @JsonInclude(JsonInclude.Include.NON_NULL)
 @JsonIgnoreProperties(ignoreUnknown = true)
+@ToString
 @SuppressWarnings("unused")
-public class LogDataResponse {
+public class RestApiResponse {
     private String msg;
     private int code;
     private int sdkErrorCode;
     private String data;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/rest/RestApiResponse.java:24]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/rest/RestApiResponse.java [hunk 20]
```diff
@@ -5,27 +5,38 @@
...
 @Data
 @NoArgsConstructor
 @AllArgsConstructor
 @JsonInclude(JsonInclude.Include.NON_NULL)
 @JsonIgnoreProperties(ignoreUnknown = true)
+@ToString
 @SuppressWarnings("unused")
-public class LogDataResponse {
+public class RestApiResponse {
     private String msg;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/rest/RestApiResponse.java:21]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/rest/RestApiResponse.java [hunk 21]
```diff
@@ -5,27 +5,38 @@
...
 import lombok.NoArgsConstructor;
+import lombok.ToString;
 
 /**
- * Encodes common fields of response received for POST logXXX requests.
+ * Encodes common fields of response received from Comet REST API.
  */
 @Data
 @NoArgsConstructor
 @AllArgsConstructor
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/rest/RestApiResponse.java:15]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/rest/RestApiResponse.java [hunk 22]
```diff
@@ -5,27 +5,38 @@
 import lombok.AllArgsConstructor;
 import lombok.Data;
 import lombok.NoArgsConstructor;
+import lombok.ToString;
 
 /**
- * Encodes common fields of response received for POST logXXX requests.
+ * Encodes common fields of response received from Comet REST API.
  */
 @Data
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/rest/RestApiResponse.java:13]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/rest/RestApiResponse.java [hunk 23]
```diff
@@ -5,27 +5,38 @@
...
 @NoArgsConstructor
 @AllArgsConstructor
 @JsonInclude(JsonInclude.Include.NON_NULL)
 @JsonIgnoreProperties(ignoreUnknown = true)
+@ToString
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/rest/RestApiResponse.java:18]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/model/FileAsset.java [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * Represents the asset downloaded from the Comet server to the local file system.
+ */
+@Data
+@AllArgsConstructor
+@ToString
+public class FileAsset {
+    Path path;
+    long size;
+    Map<String, Object> metadata;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/model/FileAsset.java:19]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/model/FileAsset.java [hunk 25]
```diff
@@ -0,0 +1,21 @@
...
+import java.util.Map;
+
+/**
+ * Represents the asset downloaded from the Comet server to the local file system.
+ */
+@Data
+@AllArgsConstructor
+@ToString
+public class FileAsset {
+    Path path;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/model/FileAsset.java:17]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/model/FileAsset.java [hunk 26]
```diff
@@ -0,0 +1,21 @@
...
+    long size;
+    Map<String, Object> metadata;
+    String assetType;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/model/FileAsset.java:20]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/model/FileAsset.java [hunk 27]
```diff
@@ -0,0 +1,21 @@
...
+import lombok.ToString;
+
+import java.nio.file.Path;
+import java.util.Map;
+
+/**
+ * Represents the asset downloaded from the Comet server to the local file system.
+ */
+@Data
+@AllArgsConstructor
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/model/FileAsset.java:14]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/model/FileAsset.java [hunk 28]
```diff
@@ -0,0 +1,21 @@
...
+import lombok.Data;
+import lombok.ToString;
+
+import java.nio.file.Path;
+import java.util.Map;
+
+/**
+ * Represents the asset downloaded from the Comet server to the local file system.
+ */
+@Data
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/model/FileAsset.java:13]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/model/FileAsset.java [hunk 29]
```diff
@@ -0,0 +1,21 @@
...
+@AllArgsConstructor
+@ToString
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/model/FileAsset.java:15]: @sonatype-lift ignore

File: comet-examples/src/main/java/ml/comet/examples/ArtifactExample.java [hunk 30]
```diff
@@ -91,7 +96,50 @@ private static void run(OnlineExperiment experiment) throws Exception {
...
+
+        // load content of the artifact asset into the memory
+        //
+        LoggedArtifactAsset asset = assets.iterator().next();
+        System.out.printf("Loading content of the artifact asset '%s' into memory\n", asset.getAssetId());
+
+        ByteBuffer assetData = asset.load();
+
+        System.out.printf(
+                "Asset's content successfully loaded into memory, data size: %d.\n\n", assetData.array().length);
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-examples/src/main/java/ml/comet/examples/ArtifactExample.java:142]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/ArtifactImpl.java [hunk 31]
```diff
@@ -35,18 +36,24 @@
  * The implementation of the {@link Artifact}.
  */
 public final class ArtifactImpl extends BaseArtifactImpl implements Artifact {
-
     @Getter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/ArtifactImpl.java:39]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/BaseArtifactImpl.java [hunk 32]
```diff
@@ -1,25 +1,57 @@
...
 import java.util.Map;
 import java.util.Set;
 
 /**
  * The base superclass for all artifact implementations.
  */
-@Data
-class BaseArtifactImpl {
+abstract class BaseArtifactImpl {
+    @Getter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/BaseArtifactImpl.java:18]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java [hunk 33]
```diff
@@ -0,0 +1,132 @@
...
+
+    private final Logger logger = LoggerFactory.getLogger(LoggedArtifactAsset.class);
+
+    @Getter
+    @Setter
+    private boolean remote;
+    @Getter
+    @Setter
+    private String assetType;
+    @Getter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java:39]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/ArtifactImpl.java (hunk 31 above)
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/ArtifactImpl.java:39]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java [hunk 34]
```diff
@@ -0,0 +1,132 @@
...
+ */
+@ToString
+public final class LoggedArtifactAssetImpl implements LoggedArtifactAsset {
+
+    private final Logger logger = LoggerFactory.getLogger(LoggedArtifactAsset.class);
+
+    @Getter
+    @Setter
+    private boolean remote;
+    @Getter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java:36]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java [hunk 35]
```diff
@@ -0,0 +1,132 @@
...
+    @Setter
+    private String assetType;
+    @Getter
+    @Setter
+    private String assetId;
+    @Setter
+    private String metadataJson;
+    @Setter
+    @Getter
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java:45]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactAssetImpl.java:33]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/LoggedArtifactImpl.java:40]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/DownloadArtifactAssetOptions.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/DownloadArtifactAssetOptions.java:16]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/http/Connection.java [hunk 36]
```diff
@@ -208,6 +213,27 @@ public Connection(@NonNull String cometBaseUrl, @NonNull String apiKey,
...
+     * @param params   the map with request parameters.
+     * @return the {@link ListenableFuture} which can be used to monitor status of the request execution.
+     */
+    public ListenableFuture<Response> downloadAsync(@NonNull File file,
+                                                    @NonNull String endpoint,
+                                                    @NonNull Map<QueryParamName, String> params) {
+        Request request = createGetRequest(this.buildCometUrl(endpoint), params);
+
+        try {
+            return this.executeDownloadAsync(request, new AsyncFileDownloadHandler(file, this.logger));
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/http/Connection.java:240]: Fixed

File: comet-java-client/src/main/java/ml/comet/experiment/impl/http/Connection.java [hunk 37]
```diff
@@ -208,6 +223,31 @@ public Connection(@NonNull String cometBaseUrl, @NonNull String apiKey,
...
+        AsyncFileDownloadHandler handler = new AsyncFileDownloadHandler(file, this.logger);
+        try {
+            handler.open();
+        } catch (Throwable e) {
+            this.logger.error("Failed to start download to the file {}", file.getPath(), e);
+            // make sure to release resources
+            handler.close();
+            return new ListenableFuture.CompletedFailure<>(e);
+        }
+        return this.executeDownloadAsync(request, handler);
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/http/Connection.java:248]: @sonatype-lift ignore

**Review by Lothiraldan — APPROVED** [review]
Looks good to me, good job!

**Review by Nimrod007 — APPROVED** [review]

**Pull request merged**